File: app/src/main/java/com/commit451/gitlab/activity/RoutingActivity.java
package com.commit451.gitlab.activity;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.widget.Toast;

import com.commit451.gitlab.R;
import com.commit451.gitlab.api.GitLabClient;
import com.commit451.gitlab.model.Account;
import com.commit451.gitlab.navigation.DeepLinker;
import com.commit451.gitlab.navigation.Navigator;
import com.commit451.gitlab.navigation.RoutingNavigator;
import com.commit451.gitlab.navigation.RoutingRouter;
import com.commit451.gitlab.util.IntentUtil;
import com.novoda.simplechromecustomtabs.SimpleChromeCustomTabs;

import timber.log.Timber;

/**
 * The easy way to do deep links. Just route everything here, and it does all the work.
 */
public class RoutingActivity extends Activity {

    RoutingRouter mRouter;

    private final RoutingNavigator mNavigator = new RoutingNavigator() {
        @Override
        public void onRouteToIssue(String projectNamespace, String projectName, String issueIid) {
            Navigator.navigateToIssue(RoutingActivity.this, projectNamespace, projectName, issueIid);
        }

        @Override
        public void onRouteToCommit(String projectNamespace, String projectName, String commitSha) {
            startActivity(LoadSomeInfoActivity.newIntent(RoutingActivity.this, projectNamespace, projectName, commitSha));
            overridePendingTransition(R.anim.fade_in, R.anim.do_nothing);
        }

        @Override
        public void onRouteToMergeRequest(String projectNamespace, String projectName, String mergeRequestId) {
            startActivity(LoadSomeInfoActivity.newMergeRequestIntent(RoutingActivity.this, projectNamespace, projectName, mergeRequestId));
            overridePendingTransition(R.anim.fade_in, R.anim.do_nothing);
        }

        @Override
        public void onRouteToProject(String namespace, String projectId) {
            Navigator.navigateToProject(RoutingActivity.this, projectId);
        }

        @Override
        public void onRouteToBuild(String projectNamespace, String projectName, String buildNumber) {
            startActivity(LoadSomeInfoActivity.newBuildIntent(RoutingActivity.this, projectNamespace, projectName, Long.valueOf(buildNumber)));
            overridePendingTransition(R.anim.fade_in, R.anim.do_nothing);
        }

        @Override
        public void onRouteToMilestone(String projectNamespace, String projectName, String milestoneNumber) {
            startActivity(LoadSomeInfoActivity.newMilestoneIntent(RoutingActivity.this, projectNamespace, projectName, milestoneNumber));
            overridePendingTransition(R.anim.fade_in, R.anim.do_nothing);
        }

        @Override
        public void onRouteUnknown(Uri uri) {
            if (mOriginalUri != null) {
                IntentUtil.openPage(RoutingActivity.this, uri.toString());
            } else {
                Toast.makeText(RoutingActivity.this, R.string.deeplink_navigate_error, Toast.LENGTH_SHORT)
                        .show();
            }
        }
    };

    Uri mOriginalUri;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        handleIntent(getIntent());
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        handleIntent(intent);
    }

    @Override
    protected void onResume() {
        super.onResume();
        SimpleChromeCustomTabs.getInstance().connectTo(this);
    }

    @Override
    protected void onPause() {
        if (SimpleChromeCustomTabs.getInstance().isConnected()) {
            SimpleChromeCustomTabs.getInstance().disconnectFrom(this);
        }
        super.onPause();
    }

    private void handleIntent(Intent intent) {
        if (intent == null || intent.getData() == null) {
            Timber.e("No url was passed. How did that happen?");
            finish();
            return;
        }
        //If it has an original uri, this means that it is an internal deep link and we
        //can still fall back to what the original uri was and just show it
        mOriginalUri = intent.getParcelableExtra(DeepLinker.EXTRA_ORIGINAL_URI);
        Uri link = intent.getData();
        Timber.d("Received deep link %s", link);
        Timber.d("Original link was %s", mOriginalUri);

        //okay so last thing, if the user has followed a link, but the user
        //is not actually signed in, we want to direct them to signin
        if (GitLabClient.getAccount() == null && Account.getAccounts(this).isEmpty()) {
            Navigator.navigateToLogin(this);
            finish();
            return;
        }
        mRouter = new RoutingRouter(mNavigator);
        mRouter.route(link);
        finish();
    }
}

File: app/src/main/java/com/commit451/gitlab/activity/RoutingActivity.kt
package com.commit451.gitlab.activity

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.widget.Toast
import com.commit451.gitlab.App
import com.commit451.gitlab.R
import com.commit451.gitlab.data.Prefs
import com.commit451.gitlab.navigation.DeepLinker
import com.commit451.gitlab.navigation.Navigator
import com.commit451.gitlab.util.IntentUtil
import timber.log.Timber


/**
 * The easy way to do deep links. Just route everything here, and it does all the work.
 */
class RoutingActivity : BaseActivity() {

    var originalUri: Uri? = null

    val navigator = object : DeepLinker.Callbacks {
        override fun onRouteToIssue(projectNamespace: String, projectName: String, issueIid: String) {
            Timber.d("Routing to issue")
            startActivity(LoadSomeInfoActivity.newIssueIntent(this@RoutingActivity, projectNamespace, projectName, issueIid))
            overridePendingTransition(R.anim.fade_in, R.anim.do_nothing)
        }

        override fun onRouteToCommit(projectNamespace: String, projectName: String, commitSha: String) {
            Timber.d("Routing to commit")
            startActivity(LoadSomeInfoActivity.newIntent(this@RoutingActivity, projectNamespace, projectName, commitSha))
            overridePendingTransition(R.anim.fade_in, R.anim.do_nothing)
        }

        override fun onRouteToMergeRequest(projectNamespace: String, projectName: String, mergeRequestIid: String) {
            Timber.d("Routing to merge request")
            startActivity(LoadSomeInfoActivity.newMergeRequestIntent(this@RoutingActivity, projectNamespace, projectName, mergeRequestIid))
            overridePendingTransition(R.anim.fade_in, R.anim.do_nothing)
        }

        override fun onRouteToProject(projectNamespace: String, projectName: String, selection: DeepLinker.ProjectSelection) {
            Timber.d("Routing to project")
            Navigator.navigateToProject(this@RoutingActivity, projectNamespace, projectName, selection)
        }

        override fun onRouteToBuild(projectNamespace: String, projectName: String, buildNumber: String) {
            Timber.d("Routing to build")
            startActivity(LoadSomeInfoActivity.newBuildIntent(this@RoutingActivity, projectNamespace, projectName, java.lang.Long.valueOf(buildNumber)!!))
            overridePendingTransition(R.anim.fade_in, R.anim.do_nothing)
        }

        override fun onRouteToMilestone(projectNamespace: String, projectName: String, milestoneNumber: String) {
            Timber.d("Routing to milestone")
            startActivity(LoadSomeInfoActivity.newMilestoneIntent(this@RoutingActivity, projectNamespace, projectName, milestoneNumber))
            overridePendingTransition(R.anim.fade_in, R.anim.do_nothing)
        }

        override fun onRouteUnknown(url: String?) {
            Timber.d("Route unknown. Opening original Uri if it exists")
            if (url != null) {
                IntentUtil.openPage(this@RoutingActivity, url.toString(), App.get().currentAccount)
            } else {
                Toast.makeText(this@RoutingActivity, R.string.deeplink_navigate_error, Toast.LENGTH_SHORT)
                        .show()
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        handleIntent(intent)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        handleIntent(intent)
    }

    override fun hasBrowsableLinks(): Boolean {
        return true
    }

    private fun handleIntent(intent: Intent?) {
        if (intent == null || intent.data == null) {
            Timber.e("No url was passed. How did that happen?")
            finish()
            return
        }
        //If it has an original uri, this means that it is an internal deep link and we
        //can still fall back to what the original uri was and just show it
        originalUri = intent.getParcelableExtra(DeepLinker.KEY_ORIGINAL_URL)
        val link = intent.data.toString()
        Timber.d("Received deep link %s", link)
        Timber.d("Original link was %s", originalUri)

        //okay so last thing, if the user has followed a link, but the user
        //is not actually signed in, we want to direct them to signin
        if (Prefs.getAccounts().isEmpty()) {
            Navigator.navigateToLogin(this)
            finish()
            return
        }
        DeepLinker.route(link, navigator)
        finish()
    }
}

